cirq/sim/density_matrix_simulator.py

```diff
@@ -435,9 +436,16 @@ def density_matrix(self):
                 |  6  |   1    |   1    |   0    |
                 |  7  |   1    |   1    |   1    |
 
+        Args:
+            copy: If True, then the returned state is a copy of the density
+            matrix. If False, then the density matrix is not copied, potentially
+            saving memory. If one only needs to read derived parameters from the
+            density matrix and store then using False can speed up simulation by
+            eliminating a memory copy.
         """
         size = np.prod(self._qid_shape, dtype=int)
-        return np.reshape(self._density_matrix, (size, size))
+        matrix = self._density_matrix.copy() if copy else self._density_matrix
+        return np.reshape(matrix, (size, size))
```

Comment:
Reshape can also return a copy but this is is probably too unlikely to worry about.